File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/40-CombinationSumIi/40-CombinationSumIi.java
// Last updated: 8/4/2026, 7:25:05 PM
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        int n = candidates.length;
        Arrays.sort(candidates);
        List<List<Integer>>result = new ArrayList<>();
        List<Integer>ll = new ArrayList<>();
        solve(candidates , target , n , 0 , result , ll);
        return result;
    }
    public void solve(int arr[] , int target , int n , int idx , List<List<Integer>>result , List<Integer>ll){
        if(target == 0){
            result.add(new ArrayList<>(ll));
            return;
        }

        for(int i = idx ; i < n ; i++){
            if(i > idx && arr[i] == arr[i-1]){
                continue;
            }

            if(arr[i] > target) break;

            ll.add(arr[i]);

            solve(arr , target - arr[i] , n , i + 1 , result , ll);
            ll.remove(ll.size() - 1);
        }
    }
}

File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/40-CombinationSumIi/40-CombinationSumIi.cpp
// Last updated: 8/4/2026, 7:24:58 PM
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        int n = candidates.size();
        sort(candidates.begin() , candidates.end());
        vector<int>temp;
        vector<vector<int>>ans;

        solve(candidates, target , 0 , n ,temp , ans);
        return ans;
    }

    void solve(vector<int>&candidates , int target , int idx , int n , vector<int>&temp , vector<vector<int>>&ans){
        if(target == 0){
            ans.push_back(temp);
            return;
        }

        for(int i = idx ; i < n ; i++){
            if(i > idx && candidates[i] == candidates[i-1]) continue;

            if(candidates[i] > target) break;

            temp.push_back(candidates[i]);

            solve(candidates , target - candidates[i] , i+1 , n , temp , ans);

            temp.pop_back();
        }
    }
};